singleapplication_p.cpp

```diff
@@ -164,31 +169,22 @@ void SingleApplicationPrivate::genBlockServerName()
 
 bool SingleApplicationPrivate::startPrimary( uint timeout )
 {
-    // Successful creation means that no main process exists
-    // So we start a QLocalServer to listen for connections
-    QLocalServer::removeServer( blockServerName );
-    server = new QLocalServer();
+    QLocalServer::removeServer(blockServerName);
+    serverThread = new ServerThread(blockServerName, this);
 
-    // Restrict access to the socket according to the
-    // SingleApplication::Mode::User flag on User level or no restrictions
-    if( options & SingleApplication::Mode::User ){
-        server->setSocketOptions( QLocalServer::UserAccessOption );
-    } else {
-        server->setSocketOptions( QLocalServer::WorldAccessOption );
-    }
+    connect(serverThread, &ServerThread::newConnection,
+            this, &SingleApplicationPrivate::slotConnectionEstablished);
 
-    QObject::connect(
-        server,
-        &QLocalServer::newConnection,
-        this,
-        &SingleApplicationPrivate::slotConnectionEstablished
-    );
+    serverThread->start();
 
-    if( server->listen( blockServerName ) )
-        return true;
+    // Wait for the server to start (with timeout)
+    QElapsedTimer timer;
+    timer.start();
+    while (!serverThread->isRunning() && timer.elapsed() < timeout) {
+        QThread::msleep(10);
+    }
 
-    delete server;
-    return false;
+    return serverThread->isRunning();
 }
 
 bool SingleApplicationPrivate::connectToPrimary( uint timeout ){
```

Comment:
This function seems unfinished. Use or delete the code under the TODO below. Make sure that the functionality described in the code comments below is included. The sendApplicationMessage function needs to use the new MessageCoder. You should implement the corresponding receive function correctly.